### ISSUE TITLE
Support for Astro components in `.astro` files

### ISSUE THREAD

**Issue by smithbm2316** [nvim-treesitter/nvim-treesitter#1763]
First off, I just wanted to say a *HUGE* thank you to all of the maintainers of this project. Adding treesitter into my Neovim setup has elevated my daily development experience *so* much. I very much appreciate all of the hard work that has gone into this plugin and ecosystem to bring Treesitter support to Neovim!

I'm opening this issue to try and get some feedback/guidance on how to go about adding language support for [Astro components in `.astro` files](https://docs.astro.build/core-concepts/astro-components). Astro components really are basically just HTML files other than two main differences:
1. Like `jsx`, in Astro components you can use curly braces to inject Javascript into your component. The behavior as far as syntax highlighting goes is identical to `jsx`.
2. At the top of each Astro component, there will be YAML-styled frontmatter fences that define an area where any Javascript can be injected into the component. Essentially, instead of using script tags like in normal HTML:
```javascript
<script>
// javascript goes here
// in HTML normally
</script>
```
Astro frontmatter fences for writing Javascript in the component would look like:
```javascript
---
// javascript goes here
// in Astro components
---

// all HTML comes down here below
<html lang="en-US">
// more html
</html>
```

After doing some preliminary reading on Treesitter, parsers, and queries, my impression is that a custom parser for Astro is likely unnecessary, as the file is essentially just Javascript if between the `---` frontmatter fences, and `jsx` if below the set of `---` frontmatter fences. I'm just a bit confused about how to get started on supporting this. I'd really appreciate if if I could have someone point me in the direction of what I might try, any docs I should read for more info, etc. Unfortunately at the moment, Astro only officially supports custom highlighting for VS Code (as with all web dev stuff 😢), even though I was able to write a custom config to get their VS Code LSP extension working nicely with `nvim-lspconfig`. 

Thank you for your help!

**Comment by theHamsta** [comment]
@smithbm2316 Astro appears to be a slight variation of HTML. So a good starting point would be to have a look onto how grammar that extend other languages are written (JSONC extends JSON, C++ extends C, CUDA extends C++). You would need to add a rule to the HTML parsers to parser the `---` content `---` part. The HTML parser uses a custom scanner which could make this slightly more complex. It might be worth to have a look on the svelte parser to see how they are extending HTML.

I'm on vacation right now so I can only give you these pointers. I can have a more detailed look onto it next weekend.

**Comment by theHamsta** [comment]
I've made an initial draft for an Astro parser: https://github.com/theHamsta/tree-sitter-html/tree/astro

It is based on the HTML parser and it overwrites two rules of the HTML parser in grammar.js. The paired `---` with new lines in between can not be expressed well in the grammar DSL. Because of this you can't have `---` right now in Astro component or in regular text. To make a proper parser out of this draft, you should

- extend the HTML parser instead of manipulating its grammar.js. Add the HTML parser to package.json and follow the approach of my CUDA parser that extends C++ (https://github.com/theHamsta/tree-sitter-cuda/). You can do this directly based on my branch of the HTML parser which has `scanner.cc` rules already renamed to astro (external rules can't be inherited and must be copied)
- create an external rule for `---`, astro_component, `---`. This should be fairly similar to the external rules for script tags.

My draft should already be sufficient to parse most Astro-files. To get highlighting:

- create a `astro/highlights.scm` file that extends `html`
- create a `astro/injections.scm` that injects `jsx` in `astro_component_script`

**Comment by theHamsta** [comment]
Just read that the front matter can only appear at the start of a file. This could make the modification of grammar.js a simpler.

**Comment by smithbm2316** [comment]
Wow! Thank you so much for going out of your way to do that, especially on vacation. I'll see how far I can get on this tomorrow and after work some days this week, I've been able to read through some of the documentation for tree-sitter already, but have a lot to still learn. Definitely still trying to wrap my head around some of the terminology, how parsers and queries work, and how the query syntax works. I've been able to get most of the highlighting to work with the branch you created on the HTML parser which is great! I've gotten a tree-sitter project scaffolded out now and will spend some time trying to get that `grammar.js` file written up. Again, thank you!

**Comment by theHamsta** [comment]
Just tell me if you need help with anything

**Comment by yashguptaz** [comment]
There has been no update on this issue for a while.
I want to try writing the grammar for Astro.
Should I go ahead? 
I would need some good help while doing so as I have never written Treesitter grammar before.

**Comment by smithbm2316** [comment]
> There has been no update on this issue for a while. I want to try writing the grammar for Astro. Should I go ahead? I would need some good help while doing so as I have never written Treesitter grammar before.

By all means, go ahead. Sorry for the lack of response on my end. I had opened this issue right when I was in the middle of some transition at work and dropped the ball on responding to this. I spent a few hours trying to grok the basics of working with Treesitter and unfortunately wasn't really able to make much progress. I ended up just using the builtin syntax highlighting file for `typescriptreact` to highlight my Astro files, as I was able to get Astro's LSP setup quite easily and get 99% of the user experience that I was looking for while doing some work on my personal site rework. @yashguptaz It would be lovely if you'd like to take a stab at it. @theHamsta thanks for all of your guidance in this issue, your swiftness of response and the amount of details you provided were very, very much appreciated.

**Comment by ShivamJoker** [comment]
How do i get syntax highlighting or autocomplete ?

**Comment by bkyerv** [comment]
> I've made an initial draft for an Astro parser: https://github.com/theHamsta/tree-sitter-html/tree/astro
> 
> It is based on the HTML parser and it overwrites two rules of the HTML parser in grammar.js. The paired `---` with new lines in between can not be expressed well in the grammar DSL. Because of this you can't have `---` right now in Astro component and regular text may not start with `---`. To make a proper parser out of this draft, you should
> 
>     * extend the HTML parser instead of manipulating its grammar.js. Add the HTML parser to package.json and follow the approach of my CUDA parser that extends C++ (https://github.com/theHamsta/tree-sitter-cuda/blob/master/package.json#L16-L17, https://github.com/theHamsta/tree-sitter-cuda/blob/master/grammar.js). The file contains just the rules that CUDA overwrites of the C++ parser. You can do this directly based on my branch of the HTML parser which has `scanner.cc` rules already renamed to astro (external rules can't be inherited and must be copied)
> 
>     * create an external rule for `---`, astro_component, `---`. This should be fairly similar to the external rules for script tags.
> 
> 
> My draft should already be sufficient to parse most Astro-files. To get highlighting:
> 
>     * create a `astro/highlights.scm` file that extends `html` (the structure of the file should probably be similar to the svelte version)
> 
>     * create a `astro/injections.scm` that injects `jsx` in `astro_component_script`

Hello @theHamsta , thank you for draft. How can I use it with nvim? when you suggest to create a file that extends html, what should be the contents of the file? Could you please explain in more details as to how to use your parser/highlighter?

**Comment by theHamsta** [comment]
In our README section, we explain how you can add a custom parser. You add a parser for `astro` with the url of my repo. Add `branch='astro'` to the install info. See parser.lua as a reference on how the parsers are specified. You can also directly add the `astro` entry there. (I'm on my cell phone so I can't easily write the actual code here)

With extending, html I was referring to the capability of tree-sitter in grammar.js to import other grammars. What I did in the draft was just copy&paste from the html parser. Importing has the advantage that your grammar will automatically profit from future updates of the html parser.

**Comment by virchau13** [comment]
My suggestion for anyone who wants to set up Astro highlighting: just set `ft=astro` to use `tsx` syntax highlighting. This can be achieved by putting something like the following:
```lua
local ft_to_parser = require('nvim-treesitter.parsers').filetype_to_parsername
ft_to_parser.astro = "tsx"
```
and putting this in `~/.config/nvim/ftdetect/astro.vim`:
```vim
autocmd BufRead,BufEnter *.astro set filetype=astro
```

The main reason why  `tsx` is a good syntax highlighter for Astro files is that Astro's grammar literally looks like
<pre><code>---
<i>{typescript}</i>
---
<i>{tsx}</i>
</pre></code>
Basically, a TSX file with optional TS frontmatter. Tree-sitter already handles error nodes, meaning that the `---`s don't need any special meaning; everything will work out of the box. And since [Astro already has a LSP server](https://github.com/withastro/language-tools), there's no point restricting the syntax highlighting further.

**Comment by virchau13** [comment]
I decided to make it anyway.

**Comment by ShivamJoker** [comment]
Thanks a lot man. Let me know if I can sponsor you anywhere.

**Comment by virchau13** [comment]
Sorry, I realised I completely forgot to add curly brace interpolation support (e.g. `<p> { 3+3 } </p>`), so I just added it. If you installed the parser already, you should probably update nvim-treesitter and `:TSUpdate astro` after #2760 gets merged.

**Comment by RilDev** [comment]
I'm getting an error when I run the command `:TSInstall astro`:

```
E5108: Error executing lua ...er/start/nvim-treesitter/lua/nvim-treesitter/install.lua:373: Parser not available for language astro
stack traceback:
        [C]: in function 'get_parser_install_info'
        ...er/start/nvim-treesitter/lua/nvim-treesitter/install.lua:373: in function 'install_lang'
        ...er/start/nvim-treesitter/lua/nvim-treesitter/install.lua:422: in function 'run'
        [string ":lua"]:1: in main chunk
```

`:TSUpdate` indicates `All parsers are up-to-date!`.

And it's not displayed in the installer info neither:

![2022-04-12_15-03](https://user-images.githubusercontent.com/13135150/162969004-1bd681a2-c578-444f-aa65-012065350030.png)

Is there anything else I have to do to get the Astro parser?

**Comment by RilDev** [comment]
Hi @theHamsta, I restarted `nvim` and it is still unavailable.

**Comment by theHamsta** [comment]
Can you check `lua/parser.lua` in your installation?

**Comment by RilDev** [comment]
I'm not sure if it's what you are asking for, but here is my `treesitter` config:

```
local status_ok, configs = pcall(require, "nvim-treesitter.configs")
if not status_ok then
  vim.notify("nvim-treesitter not found!")
  return
end

configs.setup {
  ensure_installed = "all", -- one of "all", "maintained" (parsers with maintainers), or a list of languages
  sync_install = false, -- install languages synchronously (only applied to `ensure_installed`)
  ignore_install = { "" }, -- List of parsers to ignore installing
  autopairs = {
    enable = true,
  },
  highlight = {
    enable = true, -- false will disable the whole extension
    disable = { "" }, -- list of language that will be disabled
    additional_vim_regex_highlighting = true,
  },
  indent = { enable = true, disable = { "yaml" } },
  context_commentstring = {
    enable = true,
    enable_autocmd = false,
  },
  rainbow = {
    enable = true,
    -- disable = { "jsx", "cpp" }, list of languages you want to disable the plugin for
    extended_mode = true, -- Also highlight non-bracket delimiters like html tags, boolean or table: lang -> boolean
    max_file_lines = nil, -- Do not enable for files with more than n lines, int
    -- colors = {}, -- table of hex strings
    -- termcolors = {} -- table of colour name strings
  },
}
```

I even tried to pass the `ensure_installed` to `all` to none effect.

**Comment by theHamsta** [comment]
This file on your computer: https://github.com/nvim-treesitter/nvim-treesitter/blob/master/lua/nvim-treesitter/parsers.lua Does it have astro?

**Comment by brun8** [comment]
I'm getting no syntax highlighting even after installing the parser and the lsp. When I open a .astro file the filetype option is not set, if I set it to astro manually I get some syntax highlighting but only on the jsx part. I'm on the latest versions of both neovim and treesitter, am I missing something?

**Comment by theHamsta** [comment]
@brun8 you need to add a filetype detect rule or check whether neovim already added support for astro files

**Comment by ShivamJoker** [comment]
@virchau13 I am still not sure how to get this feature to work? I have updated tree-sitter

![image](https://user-images.githubusercontent.com/23727670/172891394-7e6d51f5-f44d-4463-b6a6-a404771db228.png)

**Comment by virchau13** [comment]
@ShivamJoker You need to set the filetype to astro for nvim-treesitter to automatically detect the correct parser. This can be done automatically for `.astro` files by e.g. putting this in `~/.config/nvim/ftdetect/astro.vim`:
```vim
autocmd BufRead,BufEnter *.astro set filetype=astro
```

**Comment by ShivamJoker** [comment]
Thanks it works!

**Comment by ShivamJoker** [comment]
@virchau13 but its breaking in middle 
![image](https://user-images.githubusercontent.com/23727670/172900533-d811baf8-909c-4886-834d-745cc201d2f0.png)

also is there any formatter we can setup?

**Comment by virchau13** [comment]
> @virchau13 but its breaking in middle 

Looks like a bug. I'll try to fix it when I wake up, do you have a reproduction of the issue that I can use?

> also is there any formatter we can setup?

Maybe https://github.com/withastro/prettier-plugin- -astro? Not sure though (I've never used it)

**Comment by ShivamJoker** [comment]
You can use this file https://gist.github.com/ShivamJoker/c2c8282cbb66b5363d96164b81a7c0b2

**Comment by virchau13** [comment]
@ShivamJoker I can't reproduce the issue, I think it might be on your end

![image](https://user-images.githubusercontent.com/16955157/173169991-8673eda0-732c-4c2a-8a5c-73c9c7d165c2.png)

**Comment by ShivamJoker** [comment]
@virchau13 Can you try scrolling using the mouse?

**Comment by virchau13** [comment]
> @virchau13 Can you try scrolling using the mouse?

Still perfectly fine. Maybe you're stumbling into a bug in a release version of Neovim that's been fixed in a nightly version?

**Comment by ShivamJoker** [comment]
Ohh that might be the case. I'll wait for next release

**Comment by niftytyro** [comment]
Hi, I don't seem to be able to get syntax highlighting.
I have installed the astro tree sitter and filetype as you can see in the screenshot, is also detected. But the syntax highlighting doesn't seem to be working.
Am I missing something?

![Screenshot 2022-08-07 at 5 07 49 PM](https://user-images.githubusercontent.com/56196911/183288695-cc750845-e0c7-42e7-83dc-37aaa27bf0c9.png)

**Comment by JohnCosta27** [comment]
> Hi, I don't seem to be able to get syntax highlighting. I have installed the astro tree sitter and filetype as you can see in the screenshot, is also detected. But the syntax highlighting doesn't seem to be working. Am I missing something?
> 
> ![Screenshot 2022-08-07 at 5 07 49 PM](https://user-images.githubusercontent.com/56196911/183288695-cc750845-e0c7-42e7-83dc-37aaa27bf0c9.png)

I am having the same issue, hopefully someone can help.

**Comment by sondr3** [comment]
If you are using Neovim master/0.8.0 and above, or are on 0.7.0 and below but have enabled `filetype.lua`, the following snippet worked for me:

```lua
vim.filetype.add({
  extension = {
    astro = "astro",
  },
})
```

To enable `filetype.lua` and disable the old way of loading filetypes. Just note that this is the default now on master.

```lua
vim.g.do_filetype_lua = 1
vim.g.did_load_filetypes = 0
```

**Comment by clason** [comment]
> To enable filetype.lua and disable the old way of loading filetypes. Just note that this is the default now on master.

To be clear, that is _only_ necessary on 0.7.2 and should _not_ be done on master.

**Comment by gigo6000** [comment]
For anyone using Astro Nvim, you can add this to your `polish = function()` :
```
    -- Set up custom filetypes
    vim.filetype.add {
      extension = {
        astro = "astro",
      },
    }
```

**Comment by R4k4210** [comment]
> For anyone using Astro Nvim, you can add this to your `polish = function()` in your `~/.config/nvim/lua/user/init.lua` :
> 
> ```lua
>     -- Set up custom filetypes
>     vim.filetype.add {
>       extension = {
>         astro = "astro",
>       },
>     }
> ```

Thank! Now I only need to find a way for astro formatting x.x

**Comment by harrisoncramer** [comment]
There seems to be an issue where the parser isn't picking up interpolated values inside of the JSX expressions in the Astro template. 

<img width="795" alt="Screenshot 2023-01-27 at 8 39 38 PM" src="https://user-images.githubusercontent.com/32515581/215234978-d6ddbb89-97fc-40a1-ae3f-9398217ef2e7.png">

However, adding the following injection at `~/.config/nvim/after/queries/astro/injections.scm` fixes this for me:

```
;; extends
((interpolation
  (raw_text) @javascript)
)
```

<img width="772" alt="Screenshot 2023-01-27 at 8 40 56 PM" src="https://user-images.githubusercontent.com/32515581/215235028-aa424fa1-0d79-4331-814b-d645853e45a1.png">

**Comment by virchau13** [comment]
@harrisoncramer Have you installed the `tsx` parser? Try https://github.com/virchau13/tree-sitter-astro/#troubleshooting

**Comment by harrisoncramer** [comment]
@virchau13 Ah, missed that. Thank you, worked great.

**Comment by alvgaona** [comment]
@virchau13 is the CSS highlight issue still going on with tree-sitter? Looks like CSS isn't getting highlighted within the `<style>` tag.

**Comment by itsrobel** [comment]
I am still not getting my highlighting to work for .astro
System:
 MacOS
 Neovim` 0.10 - using kickstarter
init.lua 
```lua
  { -- Highlight, edit, and navigate code
    'nvim-treesitter/nvim-treesitter',
    event = { 'BufNewFile', 'BufReadPost' },
    build = ':TSUpdate',
    opts = {
      auto_install = true,
      sync_install = true,
      ensure_installed = {
        -- 'bash',
        -- 'c',
        -- 'diff',
        'html',
        'lua',
        -- 'luadoc',
        -- 'markdown',
        -- 'astro',
        'javascript',
        'typescript',
        'tsx',
        -- 'vim',
        -- 'vimdoc',
      },
      highlight = {
        enable = true,
        -- Some languages depend on vim's regex highlighting system (such as Ruby) for indent rules.
        --  If you are experiencing weird indenting issues, add the language to
        --  the list of additional_vim_regex_highlighting and disabled languages for indent.
        additional_vim_regex_highlighting = { 'ruby', 'astro' },
      },
      textsubjects = {
        enable = true,
      },
      -- Autoinstall languages that are not installed
      indent = { enable = true, disable = { 'ruby' } },
      incremental_selection = {
        enable = true,
        keymaps = {
          init_selection = 'gnn', -- set to `false` to disable one of the mappings
          node_incremental = 'grn',
          scope_incremental = 'grc',
          node_decremental = 'grm',
        },
      },
    },
    config = function(_, opts)
      -- [[ Configure Treesitter ]] See `:help nvim-treesitter`

      -- Prefer git instead of curl in order to improve connectivity in some environments
      -- require('nvim-treesitter.install').prefer_git = true
      ---@diagnostic disable-next-line: missing-fields
      require('nvim-treesitter.configs').setup(opts)

      -- There are additional nvim-treesitter modules that you can use to interact
      -- with nvim-treesitter. You should go explore a few and see what interests you:
      --
      --    - Incremental selection: Included, see `:help nvim-treesitter-incremental-selection-mod`
      --    - Show your current context: https://github.com/nvim-treesitter/nvim-treesitter-context
      --    - Treesitter + textobjects: https://github.com/nvim-treesitter/nvim-treesitter-textobjects
    end,
  },

```

**Comment by lucario387** [comment]
We don't support lazy loading :)
You are on your own if you're doing lazy loading

**Comment by itsrobel** [comment]
with lazy turned off/removed there is not highlighting on astro files. every other file type I use has the highlight mdx/py/tsx

**Comment by rorybyrne** [comment]
> I am still not getting my highlighting to work for .astro System: MacOS Neovim` 0.10 - using kickstarter init.lua
> 
> ```lua
>   { -- Highlight, edit, and navigate code
>     'nvim-treesitter/nvim-treesitter',
>     event = { 'BufNewFile', 'BufReadPost' },
>     build = ':TSUpdate',
>     opts = {
>       auto_install = true,
>       sync_install = true,
>       ensure_installed = {
>         -- 'bash',
>         -- 'c',
>         -- 'diff',
>         'html',
>         'lua',
>         -- 'luadoc',
>         -- 'markdown',
>         -- 'astro',
>         'javascript',
>         'typescript',
>         'tsx',
>         -- 'vim',
>         -- 'vimdoc',
>       },
>       highlight = {
>         enable = true,
>         -- Some languages depend on vim's regex highlighting system (such as Ruby) for indent rules.
>         --  If you are experiencing weird indenting issues, add the language to
>         --  the list of additional_vim_regex_highlighting and disabled languages for indent.
>         additional_vim_regex_highlighting = { 'ruby', 'astro' },
>       },
>       textsubjects = {
>         enable = true,
>       },
>       -- Autoinstall languages that are not installed
>       indent = { enable = true, disable = { 'ruby' } },
>       incremental_selection = {
>         enable = true,
>         keymaps = {
>           init_selection = 'gnn', -- set to `false` to disable one of the mappings
>           node_incremental = 'grn',
>           scope_incremental = 'grc',
>           node_decremental = 'grm',
>         },
>       },
>     },
>     config = function(_, opts)
>       -- [[ Configure Treesitter ]] See `:help nvim-treesitter`
> 
>       -- Prefer git instead of curl in order to improve connectivity in some environments
>       -- require('nvim-treesitter.install').prefer_git = true
>       ---@diagnostic disable-next-line: missing-fields
>       require('nvim-treesitter.configs').setup(opts)
> 
>       -- There are additional nvim-treesitter modules that you can use to interact
>       -- with nvim-treesitter. You should go explore a few and see what interests you:
>       --
>       --    - Incremental selection: Included, see `:help nvim-treesitter-incremental-selection-mod`
>       --    - Show your current context: https://github.com/nvim-treesitter/nvim-treesitter-context
>       --    - Treesitter + textobjects: https://github.com/nvim-treesitter/nvim-treesitter-textobjects
>     end,
>   },
> ```
> 
> ftdetect/astro.lua
> 
> ```lua
> vim.filetype.add {
>   extension = {
>     astro = 'astro',
>   },
> }
> ```
> 
> the out put of `:set filetype` is `filetype=astro` when I reinstall astro using TSInstall if the current buffer is a .astro file it will apply the highlighting but any other buffer will not be highlighted
> 
> <img alt="Screenshot 2024-07-10 at 1 13 37 AM" width="1800" src="https://private-user-images.githubusercontent.com/61813618/347318482-bd3c6569-60ab-432d-b2b1-781a4cbd0fd5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU2MzQzMjYsIm5iZiI6MTcyNTYzNDAyNiwicGF0aCI6Ii82MTgxMzYxOC8zNDczMTg0ODItYmQzYzY1NjktNjBhYi00MzJkLWIyYjEtNzgxYTRjYmQwZmQ1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTA2VDE0NDcwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTkzNWE5N2M4MDNkZTM1MDc0MWYyOWNiZDlhMzBiMTIwNzdlYWJhZjM5ZWIzYmViY2JmMGZkOGFlZTNlNDVlZWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.oBvpnFhY3OS6DJuQ1q-2khkk6YwHj2T_tllvd9xLzmY">

Same problem for me. did you ever fix it?
